### DISCUSSION TITLE
How do I know if response has already been rendered?

### DISCUSSION THREAD

**Discussion by please-fill-out-this-field** [mojolicious/mojo#2188] [Q&A] [answered]
I want this:
```perl
$app->helper('my_app.already_rendered', sub {...});

# Elsewhere
return $log->trace('Response already rendered for action "foo"')
  if $c->my_app->already_rendered;

$c->render(...);
```

I could probably just inspect `mojo.respond` or `mojo.finished` and make the decision based on that, but neither of them is documented which makes me want to avoid them:
```
$ git grep mojo.respond
lib/Mojolicious/Renderer.pm:  croak 'A response has already been rendered' if $c->stash->{'mojo.respond'}++;
$ git grep mojo.finished
lib/Mojolicious/Controller.pm:  if (!$stash->{'mojo.finished'} && ++$stash->{'mojo.finished'}) {
$
```

I assume they're both for internal use only and I'm not supposed to access them directly in application code?
Perhaps there's some "proper" way to do what I want?

**Comment by kraih** [comment]
All `mojo.*` stash values are private and will break in the future.

**Comment by rabbiveesh** [comment, marked as answer]
The safe thing to do is to see if the `$c->res` has a `code`

**Reply by please-fill-out-this-field** [reply]
Thanks, worked for me.
